File: frontend/src/api/jobDescriptions.ts
import { apiGet, apiPost } from "./client";

export interface JobDescriptionCreate {
  company_name: string;
  job_title: string;
  raw_text: string;
}

export interface JobDescriptionRead {
  id: number;
  user_id: number;
  company_name: string | null;
  job_title: string;
  raw_text: string;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface JobAnalysisRead {
  id: number;
  job_description_id: number;
  job_title: string;
  job_type: string;
  required_skills: string[];
  bonus_skills: string[];
  responsibilities: string[];
  keywords: string[];
  resume_focus_suggestions: string[];
  model_name: string;
  created_at: string;
  updated_at: string;
}

export function createJobDescription(payload: JobDescriptionCreate): Promise<JobDescriptionRead> {
  return apiPost<JobDescriptionCreate, JobDescriptionRead>("/job-descriptions", payload);
}

export function listJobDescriptions(): Promise<JobDescriptionRead[]> {
  return apiGet<JobDescriptionRead[]>("/job-descriptions");
}

export function analyzeJobDescription(jobDescriptionId: number): Promise<JobAnalysisRead> {
  return apiPost<Record<string, never>, JobAnalysisRead>(`/job-descriptions/${jobDescriptionId}/analyze`, {});
}

File: backend/app/api/routes/job_descriptions.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.ai.client import AIConfigurationError, AIResponseError
from app.core.database import get_db
from app.core.security import get_current_user
from app.models.user import User
from app.schemas.analysis import JobAnalysisRead
from app.schemas.job_description import JobDescriptionCreate, JobDescriptionRead
from app.services.job_analysis_service import JobAnalysisService, JobDescriptionNotFoundError

router = APIRouter(prefix="/job-descriptions", tags=["job descriptions"])


@router.post("", response_model=JobDescriptionRead, status_code=status.HTTP_201_CREATED)
def create_job_description(
    payload: JobDescriptionCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> JobDescriptionRead:
    service = JobAnalysisService(db)
    return service.create_job_description(payload, user_id=current_user.id)


@router.get("", response_model=list[JobDescriptionRead])
def list_job_descriptions(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> list[JobDescriptionRead]:
    service = JobAnalysisService(db)
    return service.list_job_descriptions(user_id=current_user.id)


@router.post("/{job_description_id}/analyze", response_model=JobAnalysisRead)
def analyze_job_description(
    job_description_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> JobAnalysisRead:
    service = JobAnalysisService(db)

    try:
        return service.analyze_job_description(job_description_id, user_id=current_user.id)
    except JobDescriptionNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except AIConfigurationError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc
    except AIResponseError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc)) from exc
